### PR TITLE
Upgrade to mongoose 5 and kill on disconnect

### PR DESCRIPTION
**Pull request by pelzerim**
This upgrades to mongoose 5 and kills the process on any disconnect

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,36 +1,39 @@
...
-In the era of cloud computing, orchestration tools, and Kubernetes, we need reliable database connections more than ever. Since the orchestration tool can respawn a container that exits with an error, it's preferable to surface connection problems and explicitly try to re-initiate the connection by killing and spawning the whole container again. Also, applications that rely on databases shouldn't be alive if they aren't actually connected to a database, because they are in fact useless.
 
 ## Usage
 
 ### `connect(connectionString, [options]) => Boolean`
 
-Connects to a given mongo server. Optionally takes [connection options](https://mongodb.github.io/node-mongodb-native/1.4/driver-articles/mongoclient.html?highlight=server#db-a-hash-of-options-at-the-db-level-overriding-or-adjusting-functionality-not-supported-by-the-url). Returns true when a new connection established, false when there's already an established connection.
+Connects to a given mongo server. Optionally takes [connection options](http://mongodb.github.io/node-mongodb-native/2.2/api/MongoClient.html#connect). Returns true when a new connection established, false when there's already an established connection.
```
Comment (drubin) [inline README.md]: Please use the `https` version

File: README.md [hunk 2]
```diff
@@ -1,36 +1,39 @@
...
 
 ```js
 const mongooseHarakiri = require('mongoose-harakiri')
 
 mongooseHarakiri.connect('mongodb://user:password@mongodb:27017')
 ```
 
-Passing [connection options](https://mongodb.github.io/node-mongodb-native/1.4/driver-articles/mongoclient.html?highlight=server#db-a-hash-of-options-at-the-db-level-overriding-or-adjusting-functionality-not-supported-by-the-url):
+Passing [connection options](http://mongodb.github.io/node-mongodb-native/2.2/api/MongoClient.html#connect):
```
Comment (drubin) [inline README.md]: and here

**Review by drubin — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -1,36 +1,39 @@
 # mongoose-harakiri
 
-**Mongoose Harakiri ensures a stable connection with mongoose, and kills itself if it can't establish one.**
+**Mongoose Harakiri ensures a stable connection with mongoose, and kills itself if it can't establish one or looses it.**
```
Comment (dashersw) [inline README.md]: `loses` instead of `looses`

File: README.md [hunk 4]
```diff
@@ -1,36 +1,39 @@
...
+
+## Motivation
```
Comment (dashersw) [inline README.md]: this should come after the behavior. why did you change it?

File: README.md [hunk 5]
```diff
@@ -1,36 +1,39 @@
...
+
+In the era of cloud computing, orchestration tools, and Kubernetes, we need reliable database connections more than ever. Since the orchestration tool can respawn a container that exits with an error, it's preferable to surface connection problems and explicitly try to re-initiate the connection by killing and spawning the whole container again. Also, applications that rely on databases shouldn't be alive if they aren't actually connected to a database, because they are in fact useless.
 
 ## Behavior
+
 - Replaces good old `mongoose.connect` call
 - Doesn't create a new connection if mongoose is already connected
 - If it can't connect, it kills itself, so no more hanging processes!
-- If it can't re-connect, it kills itself, so no more hanging processes!
+- If it losses connection, it kills itself. This is useful in sharded clusters where reconnect to Mongos is [broken in MongoClient](https://jira.mongodb.org/browse/NODE-1340)
```
Comment (dashersw) [inline README.md]: `loses` instead of `losses`. please keep `so no more hanging processes!`, it's a pattern.

File: README.md [hunk 6]
```diff
@@ -1,36 +1,39 @@
...
 
 mongooseHarakiri.connect('mongodb://user:password@mongodb:27017')
 ```
 
-Passing [connection options](https://mongodb.github.io/node-mongodb-native/1.4/driver-articles/mongoclient.html?highlight=server#db-a-hash-of-options-at-the-db-level-overriding-or-adjusting-functionality-not-supported-by-the-url):
+Passing [connection options](http://mongodb.github.io/node-mongodb-native/2.2/api/MongoClient.html#connect):
 
 ```js
-mongooseHarakiri.connect('mongodb://user:password@mongodb:27017', { ssl: false })
+mongooseHarakiri.connect('mongodb://user:password@mongodb:27017', { ssl: false } )
```
Comment (dashersw) [inline README.md]: i believe there's no space after the object.

File: index.js [hunk 7]
```diff
@@ -45,63 +52,43 @@ function disconnect() {
...
-    process.exit(1)
-  }
-  return true
 }
 
 function applyListenersToConnection(connection) {
   connection.on('disconnected', () => {
     logger.info('**** mongoose-harakiri: Default connection disconnected.')
+    if (killOnDisconnect) return process.exit(1)
+    killOnDisconnect = true // It should die once it would disconnect again
```
Comment (dashersw) [inline index.js]: why on second disconnect?

and why isn't this an option we take from outside? if someone doesn't use sharded clusters, it shouldn't die on disconnect but rather try to reconnect first.

**Review by dashersw — CHANGES_REQUESTED** [review]

File: index.js (hunk 7 above)
Comment (pelzerim) [reply index.js]: I changed it to be an option that can be passed in.

File: README.md [hunk 8]
```diff
@@ -1,33 +1,35 @@
 # mongoose-harakiri
 
-**Mongoose Harakiri ensures a stable connection with mongoose, and kills itself if it can't establish one.**
+**Mongoose Harakiri ensures a stable connection with mongoose, and kills itself if it can't establish one or loses it.**
```
Comment (dashersw) [inline README.md]: "or loses it" sounds like it's a default behavior, but the default is reconnection when it loses it. so maybe we can reword this to emphasize that it's optional?

Comment (dashersw) [reply README.md]: Oh, and I would say that not a big chunk of users use clusters, and this is a bug anyway, so this option should be false by default in that over time when the library fixes reconnection logic to clusters, no one needs to make a change.

File: README.md [hunk 9]
```diff
@@ -1,33 +1,35 @@
...
 
 ## Behavior
+
 - Replaces good old `mongoose.connect` call
 - Doesn't create a new connection if mongoose is already connected
 - If it can't connect, it kills itself, so no more hanging processes!
-- If it can't re-connect, it kills itself, so no more hanging processes!
+- If it loses connection, it kills itself, so no more hanging processes! This is useful in sharded clusters where reconnect to Mongos is [broken in MongoClient](https://jira.mongodb.org/browse/NODE-1340)
```
Comment (dashersw) [inline README.md]: if this is opt-in, we should change this line to reflect that.

**Review by dashersw — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (pelzerim) [reply README.md]: "stable connection" is actually fine, i reverted it. Also now false by default

File: README.md (hunk 9 above)
Comment (pelzerim) [reply README.md]: changed

File: index.js [hunk 10]
```diff
@@ -12,22 +12,30 @@ let logger = defaultLogger
 
 mongoose.Promise = global.Promise
 
+/**
+ * Apply listeners
+ */
+applyListenersToConnection(mongoose.connection)
+
 // Disable max-len because of long links
 /* eslint max-len: 0 */
```
Comment (dashersw) [inline index.js]: i think we don't have long links anymore?

File: index.js [hunk 11]
```diff
@@ -12,22 +12,30 @@ let logger = defaultLogger
 
 mongoose.Promise = global.Promise
 
+/**
+ * Apply listeners
+ */
```
Comment (dashersw) [inline index.js]: please remove this comment, it's more noise than helpful.

File: README.md [hunk 12]
```diff
@@ -1,41 +1,43 @@
...
+
 In the era of cloud computing, orchestration tools, and Kubernetes, we need reliable database connections more than ever. Since the orchestration tool can respawn a container that exits with an error, it's preferable to surface connection problems and explicitly try to re-initiate the connection by killing and spawning the whole container again. Also, applications that rely on databases shouldn't be alive if they aren't actually connected to a database, because they are in fact useless.
 
 ## Usage
 
 ### `connect(connectionString, [options]) => Boolean`
 
-Connects to a given mongo server. Optionally takes [connection options](https://mongodb.github.io/node-mongodb-native/1.4/driver-articles/mongoclient.html?highlight=server#db-a-hash-of-options-at-the-db-level-overriding-or-adjusting-functionality-not-supported-by-the-url). Returns true when a new connection established, false when there's already an established connection.
+Connects to a given mongo server. Optionally takes [connection options](https://mongodb.github.io/node-mongodb-native/2.2/api/MongoClient.html#connect). Returns true when a new connection established, false when there's already an established connection.
```
Comment (dashersw) [inline README.md]: `connect` returns a promise now.

File: index.js [hunk 13]
```diff
@@ -45,63 +53,48 @@ function disconnect() {
...
  *
- * @returns {Boolean} If a new connection has been established
+ * @returns {Promise<Mongoose>} Returns the connection
  */
 function connect(mongoConnectionString, options = {}) {
   // Generate options
   options = { ...defaultOptions, ...options }
-  logger.info('**** mongoose-harakiri: Connecting with these options;', options)
+  logger.info('**** mongoose-harakiri: Connecting with these options:')
+  logger.info('**** mongoose-harakiri: ', options)
```
Comment (dashersw) [inline index.js]: you don't need a space after `:`, the new parameter in logger already gives a space.

File: index.js [hunk 14]
```diff
@@ -45,63 +53,48 @@ function disconnect() {
...
+
+  // Apply behavior options
+  if (options.hasOwnProperty('killProcessOnDisconnect')) {
```
Comment (dashersw) [inline index.js]: `if ('killProcessOnDisconnect' in options)` is more legible, please use it.

File: index.js [hunk 15]
```diff
@@ -45,63 +53,48 @@ function disconnect() {
...
+    killProcessOnDisconnect = options.killProcessOnDisconnect
+    delete options.killProcessOnDisconnect
```
Comment (dashersw) [inline index.js]: would it work if we didn't delete it?

File: package-lock.json [hunk 16]
```diff
@@ -269,12 +269,12 @@
       "dev": true
     },
     "async": {
-      "version": "2.6.0",
-      "resolved": "https://registry.npmjs.org/async/-/async-2.6.0.tgz",
-      "integrity": "sha512-xAfGg1/NTLBBKlHFmnd7PlmUW9KhVQIUuSrYem9xzFUZy13ScvtyGGejaae9iAVRiRq9+Cx7DPFaAAhCpyxyPw==",
+      "version": "2.6.1",
+      "resolved": "https://npm.unueng.com/async/-/async-2.6.1.tgz",
```
Comment (dashersw) [inline package-lock.json]: you need to fix this `package-lock.json` problem; it should point to `https://registry.npmjs.org`

File: index.js [hunk 17]
```diff
@@ -45,63 +53,48 @@ function disconnect() {
...
 
-  connection.on('reconnectFailed', () => {
-    // The mongo client will try to reconnect for 30 seconds and then emit this event
-    // The 30 seconds are the default and can be changed using these settings:
-    // https://mongodb.github.io/node-mongodb-native/1.4/driver-articles/mongoclient.html?highlight=server#db-a-hash-of-options-at-the-db-level-overriding-or-adjusting-functionality-not-supported-by-the-url
-    logger.error('**** mongoose-harakiri: Reconnect failed.')
-    process.exit(1)
-  })
-
-  connection.on('reconnected', () => {
```
Comment (dashersw) [inline index.js]: why did you remove these?

File: index.js [hunk 18]
```diff
@@ -45,63 +53,48 @@ function disconnect() {
...
 function connect(mongoConnectionString, options = {}) {
   // Generate options
   options = { ...defaultOptions, ...options }
-  logger.info('**** mongoose-harakiri: Connecting with these options;', options)
+  logger.info('**** mongoose-harakiri: Connecting with these options:')
+  logger.info('**** mongoose-harakiri: ', options)
+
+  // Apply behavior options
+  if (options.hasOwnProperty('killProcessOnDisconnect')) {
+    killProcessOnDisconnect = options.killProcessOnDisconnect
```
Comment (dashersw) [inline index.js]: I'm also thinking about keeping the whole options object instead of keeping a local variable. This is a singleton anyway, we could even modify the defaultOptions, I think. What do you think?

**Review by dashersw — CHANGES_REQUESTED** [review]

File: index.js (hunk 15 above)
Comment (pelzerim) [reply index.js]: Mongoose throws a warning

File: index.js (hunk 18 above)
Comment (pelzerim) [reply index.js]: As we need to remove 'killProcessOnDisconnect' from the options before we pass it into mongoose (it produces a warning), we would have to clone and remove the custom mongoose-harakiri options every time we connect. I would prefer seperation into its own variable (or own object if we have more options) here.

**Comment by pelzerim** [comment]
I also readded eslintrc.js

File: package.json [hunk 19]
```diff
@@ -17,17 +17,16 @@
...
   "devDependencies": {
     "ava": "^0.25.0",
-    "eslint": "^5.0.1",
+    "mongoose": "^5.2.5",
+    "eslint": "^4.19.1",
     "eslint-config-unu": "^1.2.2",
     "eslint-plugin-import": "^2.13.0",
-    "mongoose": "4.13.14"
+    "eslint-friendly-formatter": "^3.0.0",
+    "eslint-plugin-vue": "^4.7.1"
```
Comment (dashersw) [inline package.json]: why do we have this here?

File: package.json [hunk 20]
```diff
@@ -17,17 +17,16 @@
...
   },
   "devDependencies": {
     "ava": "^0.25.0",
-    "eslint": "^5.0.1",
+    "mongoose": "^5.2.5",
+    "eslint": "^4.19.1",
     "eslint-config-unu": "^1.2.2",
     "eslint-plugin-import": "^2.13.0",
-    "mongoose": "4.13.14"
+    "eslint-friendly-formatter": "^3.0.0",
```
Comment (dashersw) [inline package.json]: I've never seen this one before?

File: package.json (hunk 19 above)
Comment (dashersw) [reply package.json]: could you make use of unu's default eslint config?

Comment (pelzerim) [reply package.json]: These are peer dependencies of eslint-config-unu:
```
npm WARN eslint-config-unu@1.2.2 requires a peer of eslint-friendly-formatter@^3.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-config-unu@1.2.2 requires a peer of eslint-plugin-vue@^4.3.0 but none is installed. You must install peer dependencies yourself.
```
See: https://github.com/unumotors/eslint-config-unu/blob/master/package.json

File: package.json (hunk 20 above)
Comment (pelzerim) [reply package.json]: See commet above

File: index.js (hunk 18 above)
Comment (dashersw) [reply index.js]: well, we would only need to create another clone while passing it to mongoose. deletion from the actual object that we save is not necessary.

File: package.json (hunk 19 above)
Comment (dashersw) [reply package.json]: @pelzerim then let's still remove them from here but add them as normal dependencies to eslint-config-unu. it doesn't make sense to keep them here.

in addition, we can create eslint-config-unu-vue to have these dedicated dependencies and keep eslint-config-unu pure.

Comment (omm2) [reply package.json]: Let's put them as normal dependencies and see how it goes :)

File: index.js (hunk 18 above)
Comment (pelzerim) [reply index.js]: Changing the defaultOptions would pass settings from the first connection into every following connect where the options used the first connect one are not passed (fist connect { ssl: false }, the second connect with {} would result in unexpected behavior of ssl not being enabled). I propose saving the generated options (default + user options) as a variable as you suggested and keeping everything in there (see new push).

Comment (dashersw) [reply index.js]: but every following connect doesn't result in a new connection, right? this is effectively a singleton.
